buildlogger/client.go

```diff
@@ -0,0 +1,279 @@
+	"github.com/mongodb/grip/level"
+	"github.com/mongodb/grip/message"
+	"github.com/mongodb/grip/send"
+	"github.com/pkg/errors"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+)
+
+type buildlogger struct {
+	ctx        context.Context
+	opts       *BuildloggerOptions
+	conn       *grpc.ClientConn
+	client     internal.BuildloggerClient
+	buffer     []*internal.LogLine
+	bufferSize int
+	*send.Base
+}
+
+// BuildloggerOptions support the use and creation of a Buildlogger log.
+type BuildloggerOptions struct {
+	// Unique information to identify the log.
+	Project       string
+	Version       string
+	Variant       string
+	TaskName      string
+	TaskID        string
+	Execution     int32
+	TestName      string
+	Trial         int32
+	ProcessName   string
+	LogFormatText bool
+	LogFormatJSON bool
+	LogFormatBSON bool
+	Arguments     map[string]string
+	Mainline      bool
+
+	// Configure a local sender for "fallback" operations and to collect
+	// the location of the buildlogger output.
+	Local send.Sender
+
+	// The number max number of bytes to buffer before sending log data
+	// over rpc to Cedar.
+	MaxBufferSize int
+
+	// The gRPC client connection. If nil, a new connection will be
+	// established with the gRPC connection configuration.
+	ClientConn *grpc.ClientConn
+
+	// Configuration for gRPC client connection.
+	RPCAddress string
+	Insecure   bool
+	CAFile     string
+	CertFile   string
+	KeyFile    string
+
+	logID    string
+	format   internal.LogFormat
+	exitCode int32
+}
+
+func (opts *BuildloggerOptions) validate() error {
+	count := 0
+	if opts.LogFormatText {
+		opts.format = internal.LogFormat_LOG_FORMAT_TEXT
+		count++
+	}
+	if opts.LogFormatJSON {
+		opts.format = internal.LogFormat_LOG_FORMAT_JSON
+		count++
+	}
+	if opts.LogFormatBSON {
+		opts.format = internal.LogFormat_LOG_FORMAT_BSON
+		count++
+	}
+	if count > 1 {
+		return errors.New("cannot specify more than one log format")
+	}
+
+	if opts.ClientConn == nil {
+		if opts.RPCAddress == "" {
+			return errors.New("must specify a RPC address when a client connection is not provided")
+		}
+		if !opts.Insecure && (opts.CAFile == "" || opts.CertFile == "" || opts.KeyFile == "") {
+			return errors.New("must specify credential files when making a secure connection over RPC")
+		}
+	}
+
+	if opts.Local == nil {
+		opts.Local = send.MakeNative()
+	}
+
+	if opts.MaxBufferSize == 0 {
+		// TODO: figure out ideal default size
+		opts.MaxBufferSize = 4096
+	}
+
+	return nil
+}
+
+// SetExitCode sets the exit code variable.
+func (opts *BuildloggerOptions) SetExitCode(i int32) { opts.exitCode = i }
+
+// GetLodID returns the unique buildlogger log ID set after NewBuildlogger is
+// called.
+func (opts *BuildloggerOptions) GetLogID() string {
+	return opts.logID
+}
+
+// NewBuildlogger returns a grip Sender backed by Cedar Buildlogger.
+func NewBuildlogger(ctx context.Context, name string, l send.LevelInfo, opts *BuildloggerOptions) (send.Sender, error) {
```

Comment:
There should be a MakeLogger implementation that doesn't require you to pass a level.